test/loading.jl

```diff
@@ -2,22 +2,51 @@
 
 using Base.Test
 
-@test @__LINE__ == 5
+# Tests for @__LINE__ inside and outside of macros
+@test (@__LINE__) == 6
+
+macro macro_caller_lineno()
+    @test 9 == (@__LINE__) != __source__.line > 12
+    return __source__.line
+end
+
+@test @macro_caller_lineno() == (@__LINE__) > 12
+
+# @__LINE__ in a macro expands to the location of the macrocall in the source
+# while __source__.line is the location of the macro caller
+macro nested_LINE_expansion()
+    return quote
+        return (@emit_LINE, $(__source__.line))
+    end
+end
+macro nested_LINE_expansion2()
+    return :((@emit_LINE, $(__source__.line)))
+end
+macro emit_LINE()
+    return quote
+        (@__LINE__, $(__source__.line))
+    end
+end
+@test (@emit_LINE) == ((@__LINE__) - 3, @__LINE__)
+@test @nested_LINE_expansion() == ((@__LINE__() - 4, @__LINE__() - 12), @__LINE__())
+@test @nested_LINE_expansion2() == ((@__LINE__() - 5, @__LINE__() - 9), @__LINE__())
```

Comment:
@vtjnash @ihnorton with this approach, it looks like there's no way to create a wrapper macro which should be transparent to `@LINE`, in the same way that non-block `quote` doesn't generate normal line number nodes.

In `MicroLogging` I've got `@info`, `@warn` implemented as trivial wrappers around an `@logmsg` macro which has all the guts of the logging implementation, and I was hoping to have `@LINE` expanded inside `@logmsg`.  Of course, for the particular purposes of my package I can work around this limitation, but I'd rather not have to.

Do you see a viable way to allow such wrapper macros using the `__source__` approach you've implemented here?

One probably ugly possibility would be to have access to a `Expr(:macrocall_with_source, ...)` (better name needed!) where the user can pass the `__source__` argument explicitly, bypassing the expansion step?